### PR TITLE
feat(api): allow volume input for blowout

### PR DESCRIPTION
**Pull request by caila-marashaj**
This adds an optional input parameter to the api's `blow_out` function, which allows the user to specify a volume in microliters to blow out. Using this parameter will blow out the specified volume past the pipette's bottom position.

### CODE REVIEW

File: api/src/opentrons/hardware_control/api.py [hunk 1]
```diff
@@ -962,16 +962,27 @@ async def dispense(
         else:
             dispense_spec.instr.remove_current_volume(dispense_spec.volume)
 
-    async def blow_out(self, mount: top_types.Mount) -> None:
+    async def blow_out(
+        self, mount: top_types.Mount, microliters: Optional[float] = None
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/api.py]:
```suggestion
        self, mount: top_types.Mount, volume: Optional[float] = None
```

Keeps it consistent with other functions that take volumes

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Let's make the name of the argument consistent and then we definitely need some tests with volumes set, on both ot2 and ot3, checking that
- we appropriately raise the error if you blow out too much
- spot checking, maybe with `hypothesis`, intermediate values

File: api/src/opentrons/hardware_control/api.py [hunk 2]
```diff
@@ -962,16 +962,27 @@ async def dispense(
...
-    async def blow_out(self, mount: top_types.Mount) -> None:
+    async def blow_out(
+        self, mount: top_types.Mount, volume: Optional[float] = None
+    ) -> None:
         """
         Force any remaining liquid to dispense. The liquid will be dispensed at
         the current location of pipette
         """
-        blowout_spec = self.plan_check_blow_out(mount)
+        blowout_spec = self.plan_check_blow_out(mount, microliters)
```
Comment (Laura-Danielle) [inline api/src/opentrons/hardware_control/api.py]: the `plan_check_blow_out` should be calculating the distance required to move and throwing an error if the blow out distance calculated is _greater_ than the max blowout position.

File: api/src/opentrons/hardware_control/api.py [hunk 3]
```diff
@@ -962,16 +962,27 @@ async def dispense(
...
+
+        instrument = self.get_pipette(mount)
+        max_blowout_pos = instrument.config.blow_out
+        # start at the bottom position and move additional distance det. by plan_check_blow_out
+        blowout_distance = instrument.config.bottom - blowout_spec.plunger_distance
+
+        if blowout_distance < max_blowout_pos:
```
Comment (Laura-Danielle) [inline api/src/opentrons/hardware_control/api.py:979]: `if blowout_distance > max_blowout_pos` right?

File: api/src/opentrons/hardware_control/instruments/ot2/pipette_handler.py [hunk 4]
```diff
@@ -609,25 +609,34 @@ def plan_check_dispense(  # type: ignore[no-untyped-def]
...
 
     @overload
-    def plan_check_blow_out(self, mount: OT3Mount) -> LiquidActionSpec[OT3Axis]:
+    def plan_check_blow_out(
+        self, mount: OT3Mount, microliters: Optional[float] = None
+    ) -> LiquidActionSpec[OT3Axis]:
         ...
 
-    def plan_check_blow_out(self, mount):  # type: ignore[no-untyped-def]
+    def plan_check_blow_out(self, mount, microliters: Optional[float] = None):  # type: ignore[no-untyped-def]
```
Comment (Laura-Danielle) [inline api/src/opentrons/hardware_control/instruments/ot2/pipette_handler.py]: should also be volume.

File: api/src/opentrons/hardware_control/instruments/ot2/pipette_handler.py [hunk 5]
```diff
@@ -609,25 +609,34 @@ def plan_check_dispense(  # type: ignore[no-untyped-def]
...
         """Check preconditions and calculate values for blowout."""
         instrument = self.get_pipette(mount)
         self.ready_for_tip_action(instrument, HardwareAction.BLOWOUT)
         speed = self.plunger_speed(
             instrument, instrument.blow_out_flow_rate, "dispense"
         )
+        if microliters:
```
Comment (Laura-Danielle) [inline api/src/opentrons/hardware_control/instruments/ot2/pipette_handler.py]: can you put a TODO about this following the flex logic once the ticket I linked you before is done?

File: shared-data/python/opentrons_shared_data/pipette/dev_types.py [hunk 6]
```diff
@@ -133,6 +133,8 @@ class PipetteNameSpec(TypedDict):
     defaultAspirateFlowRate: PipetteConfigElementWithPerApiLevelValue
     defaultDispenseFlowRate: PipetteConfigElementWithPerApiLevelValue
     defaultBlowOutFlowRate: PipetteConfigElementWithPerApiLevelValue
+    # defaultBlowOutVolume: float
```
Comment (Laura-Danielle) [inline shared-data/python/opentrons_shared_data/pipette/dev_types.py]: Accidentally commented out? Do these need to be moved somewhere else?

**Review by Laura-Danielle — COMMENTED** [review]
Agree with all of seth's comments. Have a few additional comments as well. In general, looking good!

File: api/src/opentrons/hardware_control/instruments/ot2/pipette_handler.py (hunk 5 above)
Comment (SyntaxColoring) [inline api/src/opentrons/hardware_control/instruments/ot2/pipette_handler.py]: Should this be `if microliters is None`? Right now, it looks like this will use the default value if you specify `microliters=0`, which I would find surprising.

File: api/src/opentrons/hardware_control/instruments/ot3/pipette_handler.py [hunk 7]
```diff
@@ -573,18 +576,28 @@ def plan_check_dispense(
...
         speed = self.plunger_speed(
-            instrument, instrument.blow_out_flow_rate, "dispense"
+            instrument,
+            instrument.blow_out_flow_rate,
+            "blowout",
         )
 
+        ul = (
+            microliters
+            or self.get_attached_instrument(mount)["default_blow_out_volume"]
```
Comment (SyntaxColoring) [inline api/src/opentrons/hardware_control/instruments/ot3/pipette_handler.py]: Same deal with `0` being just as falsey as `None`.

File: api/tests/opentrons/hardware_control/test_ot3_api.py [hunk 8]
```diff
@@ -215,6 +215,17 @@ async def mock_refresh(ot3_hardware: ThreadManager[OT3API]) -> Iterator[AsyncMoc
     ) as mock_refresh:
         yield mock_refresh
 
+# @pytest.fixture
+# async def mock_pick_up_tip(ot3_hardware: ThreadManager[OT3API]) -> Iterator[AsyncMock]:
```
Comment (sfoster1) [inline api/tests/opentrons/hardware_control/test_ot3_api.py]: need to uncomment

File: api/tests/opentrons/hardware_control/test_ot3_api.py [hunk 9]
```diff
@@ -373,6 +384,37 @@ async def _update_position(
...
+    "load_configs",
+    [
+        {OT3Mount.LEFT: {"channels": 1, "version": (3, 3), "model": "p1000"}},
+        {OT3Mount.RIGHT: {"channels": 8, "version": (3, 3), "model": "p50"}},
+        {OT3Mount.LEFT: {"channels": 96, "model": "p1000", "version": (3, 3)}},
+    ]
+)
+async def test_blow_out(
+        ot3_hardware: ThreadManager[OT3API],
+        load_configs: Dict[str, Union[int, str, Tuple[int, int]]],
```
Comment (sfoster1) [inline api/tests/opentrons/hardware_control/test_ot3_api.py]: might be a little more readable if we make these dataclasses or something

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Looking good! Agree with laura and max, got this last bit of commented code still

File: api/src/opentrons/hardware_control/api.py [hunk 10]
```diff
@@ -962,16 +962,27 @@ async def dispense(
...
-    async def blow_out(self, mount: top_types.Mount) -> None:
+    async def blow_out(
+        self, mount: top_types.Mount, microliters: Optional[float] = None
+    ) -> None:
         """
         Force any remaining liquid to dispense. The liquid will be dispensed at
         the current location of pipette
         """
-        blowout_spec = self.plan_check_blow_out(mount)
+        blowout_spec = self.plan_check_blow_out(mount, microliters)
```
Comment (Laura-Danielle) [inline api/src/opentrons/hardware_control/api.py]: This should be 'volume' not 'microliters'

File: api/src/opentrons/hardware_control/api.py (hunk 1 above)
Comment (Laura-Danielle) [inline api/src/opentrons/hardware_control/api.py]: This should be 'volume' not 'microliters'

File: api/src/opentrons/hardware_control/protocols/liquid_handler.py [hunk 11]
```diff
@@ -81,7 +81,7 @@ async def dispense(
         """
         ...
 
-    async def blow_out(self, mount: Mount) -> None:
+    async def blow_out(self, mount: Mount, microliters: Optional[float] = None) -> None:
```
Comment (Laura-Danielle) [inline api/src/opentrons/hardware_control/protocols/liquid_handler.py]: this should be 'volume' not 'microliters'

File: api/tests/opentrons/hardware_control/test_ot3_api.py [hunk 12]
```diff
@@ -384,6 +385,104 @@ async def _update_position(
...
+            < instr_data["config"].plunger_positions_configurations.blow_out
+        )
+        # make sure calculated position is roughly what we expect
+        assert position_result[pipette_axis] == pytest.approx(
+            expected_position, rel=0.1
+        )
+
+
+@pytest.mark.parametrize("load_configs", load_blowout_configs)
+@given(blowout_volume=strategies.floats(min_value=1, max_value=200))
```
Comment (Laura-Danielle) [inline api/tests/opentrons/hardware_control/test_ot3_api.py]: Curious how much this slows down this test suite. I know that using strategies.<whatever> usually generates around 100 elements by default.

We should specify `@settings(max_examples=10)` or something similar so that we don't re-run this test so many times. If there are certain edge cases you definitely want to test, then you can specify with: `@example(<value>)`.

**Review by Laura-Danielle — CHANGES_REQUESTED** [review]
couple small changes, but this is looking really good. Nice job w the hypothesis test.

File: api/src/opentrons/hardware_control/ot3api.py [hunk 13]
```diff
@@ -1474,19 +1474,37 @@ async def dispense(
...
         Force any remaining liquid to dispense. The liquid will be dispensed at
         the current location of pipette
         """
         realmount = OT3Mount.from_mount(mount)
-        blowout_spec = self._pipette_handler.plan_check_blow_out(realmount)
+        instrument = self._pipette_handler.get_pipette(realmount)
+        blowout_spec = self._pipette_handler.plan_check_blow_out(realmount, volume)
+
+        max_blowout_pos = instrument.plunger_positions.blow_out
+        # start at the bottom position and move additional distance det. by plan_check_blow_out
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/ot3api.py]: nit: just use two lines instead of abbreviating

File: api/src/opentrons/hardware_control/ot3api.py [hunk 14]
```diff
@@ -1474,19 +1474,37 @@ async def dispense(
...
+        blowout_distance = (
+            instrument.plunger_positions.bottom + blowout_spec.plunger_distance
+        )
+        if blowout_distance > max_blowout_pos:
+            raise ValueError(
+                f"Blow out distance exceeds plunger position limit\n blowout dist = {blowout_distance}"
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/ot3api.py]: nit: don't put newline literals in here

**Review by sfoster1 — APPROVED** [review]
This is pretty awesome, I love to have this kind of testing in hardware.

**Review by Laura-Danielle — APPROVED** [review]
lgtm

File: api/Pipfile [hunk 15]
```diff
@@ -41,3 +41,4 @@ typing-extensions = ">=4.0.0,<5"
 pytest-profiling = "~=1.7.0"
 # TODO(mc, 2022-03-31): upgrade sphinx, remove this subdep pin
 jinja2 = ">=2.3,<3.1"
+hypothesis = "~=6.36.1"
```
Comment (SyntaxColoring) [inline api/Pipfile]: Oooh, very excited to try this.

Note that the version specifier `~=6.36.1` pins us to the `.36` minor version of Hypothesis, [equivalent to](https://peps.python.org/pep-0440/#compatible-release) `>=6.36.1, <6.37`. (I hate that `~=` works like this. It's really surprising and annoying.)

Since Hypothesis [uses semver](https://hypothesis.readthedocs.io/en/latest/supported.html), I think we should pin to the major version instead, by writing either of these:

* `~=6.36`
* `>=6.36,<7` (my personal soft preference because `~=` is confusing)
